x-pack/filebeat/input/cometd/input.go

```diff
@@ -0,0 +1,196 @@
+	"github.com/elastic/beats/v7/libbeat/beat"
+	"github.com/elastic/beats/v7/libbeat/common"
+	"github.com/elastic/beats/v7/libbeat/common/atomic"
+	"github.com/elastic/beats/v7/libbeat/logp"
+
+	bay "github.com/elastic/bayeux"
+)
+
+const (
+	inputName = "cometd"
+)
+
+// Run starts the input worker then returns. Only the first invocation
+// will ever start the worker.
+func (in *cometdInput) Run() {
+	var err error
+	in.workerOnce.Do(func() {
+		in.workerWg.Add(1)
+		go func() {
+			in.log.Info("Input worker has started.")
+			defer in.log.Info("Input worker has stopped.")
+			defer in.workerWg.Done()
+			defer in.workerCancel()
+			in.b = bay.Bayeux{}
+			in.creds, err = bay.GetSalesforceCredentials(in.authParams)
+			if err != nil {
+				in.log.Error("not able to get access token", err)
+				return
+			}
+			if err := in.run(); err != nil {
+				in.log.Error(err)
+				return
+			}
+		}()
+	})
+}
+
+func (in *cometdInput) run() error {
+	in.out = in.b.Channel(in.out, "-1", *in.creds, in.config.ChannelName)
+
+	var event Event
+	for e := range in.out {
+		if !e.Successful {
+			// To handle the last response where the object received was empty
+			if e.Data.Payload == nil {
+				return nil
+			}
+
+			// Convert json.RawMessage response to []byte
+			msg, err := e.Data.Payload.MarshalJSON()
+			if err != nil {
+				return fmt.Errorf("JSON error: %w", err)
+			}
+
+			// Extract event IDs from json.RawMessage
+			err = json.Unmarshal(e.Data.Payload, &event)
+			if err != nil {
+				return fmt.Errorf("error while parsing JSON: %w", err)
+			}
+			if ok := in.outlet.OnEvent(makeEvent(event.EventId, string(msg))); !ok {
+				in.log.Debug("OnEvent returned false. Stopping input worker.")
+				close(in.out)
+				return fmt.Errorf("error ingesting data to elasticsearch")
+			}
+		}
+	}
+	return nil
+}
+
+func init() {
+	err := input.Register(inputName, NewInput)
+	if err != nil {
+		panic(fmt.Errorf("failed to register %v input: %w", inputName, err))
+	}
+}
+
+// NewInput creates a new CometD input that consumes events from
+// a topic subscription.
+func NewInput(
+	cfg *common.Config,
+	connector channel.Connector,
+	inputContext input.Context,
+) (inp input.Input, err error) {
+	// Extract and validate the input's configuration.
+	conf := defaultConfig()
+	if err = cfg.Unpack(&conf); err != nil {
+		return nil, err
+	}
+
+	var authParams bay.AuthenticationParameters
+	authParams.ClientID = conf.Auth.OAuth2.ClientID
+	authParams.ClientSecret = conf.Auth.OAuth2.ClientSecret
+	authParams.Username = conf.Auth.OAuth2.User
+	authParams.Password = conf.Auth.OAuth2.Password
+	authParams.TokenURL = conf.Auth.OAuth2.TokenURL
+
+	logger := logp.NewLogger("cometd").With(
+		"pubsub_channel", conf.ChannelName)
+
+	// Wrap input.Context's Done channel with a context.Context. This goroutine
+	// stops with the parent closes the Done channel.
+	inputCtx, cancelInputCtx := context.WithCancel(context.Background())
+	defer cancelInputCtx()
```

Comment:
I'm not sure this is right, `inputCtx` will be cancelled immediately when this function is called.

What you had before was more correct:

```go
	// Wrap input.Context's Done channel with a context.Context. This goroutine
	// stops with the parent closes the Done channel.
	inputCtx, cancelInputCtx := context.WithCancel(context.Background())
	go func() {
		defer cancelInputCtx()
		select {
		case <-inputContext.Done:
		case <-inputCtx.Done():
		}
	}()
```

Which you could simplify to this:

```go
	// Wrap input.Context's Done channel with a context.Context. This goroutine
	// stops with the parent closes the Done channel.
	inputCtx, cancelInputCtx := context.WithCancel(context.Background())
	go func() {
		<-inputContext.Done
        cancelInputCtx()
	}()
```

Reply:
I don't see `workerCtx` used anywhere which would explain why this didn't cause problems in tests. Is the input not cancellable?